### PR TITLE
Basic stack pooling allocator

### LINKED ISSUE #26: Guard pages for fiber stacks
Our custom `malloc` for fiber stacks do not include guard pages, so a stack overflow may result in a corrupted program.

### LINKED ISSUE #91: Pooling allocator for stacks
We should a pooling allocation strategy for stacks to amortise their allocation costs.

### PR DESCRIPTION
**Pull request by dhil**
This patch adds a basic stack pooling allocator for WasmFX stacks. It is gated behind a feature flag `wasmfx_pooling_allocator`, which is disabled by default. ~~Currently, the pooling allocator works only with the baseline implementation as the optimized implementation segfaults (Frank, I probably need your help to figure out why)~~. By default stacks are safe, meaning they have a page guard attached. The old unsafe stacks are gated behind the feature flag `unsafe_wasmfx_stacks`.

In addition, this patch also adds missing WasmFX features to `wasmtime-c-api` and `wasmtime-c-api-impl`.

Merge after #168.

Resolves #91. Resolves #26.

### CODE REVIEW

File: crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+    if #[cfg(feature = "wasmfx_baseline")] {
+        use wasmtime_fiber as fiber_impl;
+    } else {
+        use crate::vm::fibre as fiber_impl;
+    }
+}
+
+pub type FiberStack = fiber_impl::FiberStack;
+
+#[cfg(not(feature = "wasmfx_pooling_allocator"))]
```
Comment (frank-emrich) [inline crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs:20]: From a quick test, it seems like things still compile with and without `--features wasmfx_pooling_allocator` if this module as well as `wasmfx_pooling` are enabled unconditionally.

File: crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+    use crate::runtime::vm::instance::allocator::pooling::{
+        index_allocator::{SimpleIndexAllocator, SlotId},
+        round_up_to_pow2,
+    };
+    use crate::runtime::vm::sys::vm::{commit_pages, decommit_pages};
+    use crate::vm::Mmap;
+    use anyhow::{anyhow, bail, Context, Result};
+
+    /// Represents a pool of execution stacks (used for the async fiber implementation).
```
Comment (frank-emrich) [inline crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs:64]: Used for the wasmfx implementation! :) The sentence below should probably be adapted to reflect our usage of stacks more concretely.

File: crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+            let stack_size = self.stack_size - self.page_size;
+            let bottom_of_stack = top - stack_size;
+            let start_of_stack = bottom_of_stack - self.page_size;
+            assert!(start_of_stack >= base && start_of_stack < (base + len));
+            assert!((start_of_stack - base) % self.stack_size == 0);
+
+            let index = (start_of_stack - base) / self.stack_size;
+            assert!(index < self.max_stacks);
+
+            if self.async_stack_zeroing {
```
Comment (frank-emrich) [inline crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs:198]: Hm, it seems a bit weird to re-use `async`s config options here. How about dropping the zero-ing support for now until we add our own corresponding flags?

File: crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs [hunk 4]
```diff
@@ -0,0 +1,261 @@
+// WasmFX fiber stack allocators
+//
+// * on_demand: allocates memory lazily
+// * pooling: preallocates a chunk of memory eagerly
+//
+
+use anyhow::Result;
+use wasmtime_continuations::WasmFXConfig;
+
+cfg_if::cfg_if! {
```
Comment (frank-emrich) [inline crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs:10]: How about making each wasmfx implementation export the `Fiber` type it uses?

**Review by frank-emrich — APPROVED** [review]
LGTM with some comments

File: crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs (hunk 4 above)
Comment (dhil) [reply crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs:10]: Yeah, we should do something like that.

File: crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs (hunk 3 above)
Comment (dhil) [reply crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs:198]: Can do. I plan to experiment with options like these soon.

File: crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs (hunk 2 above)
Comment (dhil) [reply crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs:64]: Yes.

File: crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs (hunk 1 above)
Comment (dhil) [reply crates/wasmtime/src/runtime/vm/instance/allocator/wasmfx_allocator.rs]: Undone. It is necessary to pass the `nostd` checks.

**Pull request merged**